script/babel-transform-constants.mjs

```diff
@@ -1,27 +1,62 @@
-module.exports = transform
+export default transform
 
-var path = require('path')
-var resolveFrom = require('resolve-from')
+import path from 'path'
+import resolveFrom from 'resolve-from'
+import codes from '../lib/character/codes.js'
+import values from '../lib/character/values.js'
+import constants from '../lib/constant/constants.js'
+import types from '../lib/constant/types.js'
 
 var supported = [
-  require.resolve('../lib/character/codes'),
-  require.resolve('../lib/character/values'),
-  require.resolve('../lib/constant/constants'),
-  require.resolve('../lib/constant/types')
+  'micromark/lib/character/codes.js',
+  'micromark/lib/character/values.js',
+  'micromark/lib/constant/constants.js',
+  'micromark/lib/constant/types.js'
```

Comment:
The reason for this handling is because of: https://github.com/syntax-tree/mdast-util-from-markdown/issues/4, which landed in https://github.com/micromark/micromark/commit/bf53bf93074068376240dc692267f1082325bfd4.

I’m assuming your change works fine in this project (although: windows?), but I’m wondering if it’d also work on other projects using this script and having the files in `node_modules/micromark/...`?

Reply:
Btw, I always thought windows just works with direct slashes. Backslash is just default.

Reply:
I don't get from the commit. Where micromark is in mdast-util-from-markdown?

Reply:
These strings are not used for importing/requiring/reading so safe to use in windows.

Reply:
Node handles `/` in requires, but not in path related stuff.

For the commits/prs I references: it’s indeed rather complex. `mdast-util-from-markdown` is using the micromark parser, but adding a different “compiler”: it doesn’t go to HTML but goes to an AST. This is not documented (yet) — most folks won’t do it, but it is powerful and intended to be used. To make writing `from-markdown` easier, I’m also using constants, types, etc there. But: to make it fast, I also want to compile them away: https://github.com/syntax-tree/mdast-util-from-markdown/commit/e59424359047d9ee47b83f090665f93d2c23349e#diff-92bbac9a308cd5fcf9db165841f2d90ce981baddcb2b1e26cfff170929af3bd1R5-R8

Reply:
I don't understand how this works. How micromark happens in `./micromark`?
```
babel lib/ --out-dir dist/ --quiet --retain-lines --plugins ./micromark/script/babel-transform-constants.js; prettier dist/ --loglevel error --write
```

Reply:
it’s cloned in from-markdown: https://github.com/syntax-tree/mdast-util-from-markdown/blob/264b0c015db1835470cb42ed66b0a1564e837c4a/package.json#L63

A bit hacky, totally 😅

Reply:
Hm... I would protect it with tag version

Reply:
Sorry, I think I’ve explained this rather rather badly.

I don’t think it needs to be protected. I think the problem is: two projects use `babel-transform-constants.mjs`:

- micromark itself, in which case it should strip:
  - `micromark/lib/character/codes.js`
  - `micromark/lib/character/values.js`
  - ...
- and `mdast-util-from-markdown`, in which case it should strip:
  - `mdast-util-from-markdown/node_modules/micromark/lib/character/codes.js`
  - `mdast-util-from-markdown/node_modules/micromark/lib/character/values.js`
  - ...

Hence, the code here used `require.resolve`.

The fact that `mdast-util-from-markdown` uses git to get the babel scripts, and whether it’s git clone or a submodule, shouldn’t matter. A different approach to get it in from-markdown would be to `curl https://raw.githubusercontent.com/micromark/micromark/main/script/babel-transform-constants.js > script/babel-transform-constants.js`.

Reply:
Maybe easier to keep both project in the same repository? Babel script problem will go away. And it would be easier to  upgrade from-markdown on micromark changes.

Reply:
That is an unrelated problem. If they were in a monorepo, then lerna or whatever would put `repo/packages/mdast-util-from-markdown/node_modules/micromark` in from-markdown.

I don’t think you need to change anything in from-markdown. The problem is in your change here, where you removed `require.resolve`. What is the reason `require.resolve` cannot be used?

Reply:
Es modules does not have require

Reply:
Btw, yarn monorepo links all packages. So `repo/packages/mdast-util-from-markdown/node_modules/micromark` === `repo/packages/micromark`. NPM 7 probably does the same. Though there would be easier to setup script in the root of the project. And easier to check if it works. And from-markdown build step would not be immediately broken after merging this PR.

Reply:
Ah, right, ok let me debug this some more.

One thought: could the path from `import.meta` be used with `resolve-from`?

Reply:
There is a function coming in Node, but on latest Node it’s still under a flag: https://nodejs.org/api/esm.html#esm_no_require_resolve

Reply:
Quick Q if you have the time: Can babel support ESM plugins? `babel ... --plugins ./micromark/script/babel-transform-constants.mjs` crashes

Reply:
Not yet. Babel 8 will.

Reply:
Hm, I missed import.meta.url is supported in node 10. Will try to make it work.

Reply:
One problem is createRequireFromPath does not accept import.meta.url. createRequire which accepts is supported only since node 12.

Reply:
Oh, this solves the problem
https://nodejs.org/docs/latest-v10.x/api/url.html#url_url_fileurltopath_url

Reply:
Ohh yeah, I~’m now looking at your new rollup config and was indeed thinking `createRequire` might work

Reply:
It works. I copy pasted rollup config to from-markdown and all works with the last commit.

Reply:
Ah, shit, does not work in dubnium because of rollup.
Anyway this will be fixed when all lib will migrate to modules.
Btw I already migrated everything locally. Is it ok to merge as is?

Reply:
I reverted latest commit

Reply:
As in, from-markdown fails on dubnium?

Reply:
Both fails yet with import.meta.url so I reverted to hardcoded path which does not work in from-markdown. I cannot fix it in this PR but will do this in the next with migrating lib to es modules.